clamav/cnudie.py

```
@@ -23,20 +28,105 @@
+
+        if isinstance(previous_finding.data, dict):
+            # In this case the previous scan we have available predates the current dataclasses.
+            # It is safe to assume that this scan result is very old.
+            return False
+
+        clamav_scan_metadata = previous_finding.data.metadata
+        if (
+            skip_if_virus_definition_unchanged
+            and clamav_scan_metadata.signature_version == current_version_info.signature_version
+        ):
+            return True
+
+        if (
+            current_version_info.signature_date
+            - clamav_scan_metadata.virus_definition_timestamp
+        ).days < virus_definition_age_threshold:
+            return True
+
+        return False
 
     def scan_resource(
         resource_node: cnudie.iter.ResourceNode,
-    ):
+    ) -> clamav.model.ClamAVResourceScanResult:
         component = resource_node.component
         resource = resource_node.resource
 
+        previous_findings = delivery.util.artefact_metadata_for_resource_node(
+            delivery_client=delivery_client,
+            resource_node=resource_node,
+            type=dso.model.Datatype.MALWARE,
+        ) if delivery_client else []
+
+        sorted_findings = sorted(
+            previous_findings,
+            key=lambda f: f.meta.creation_date.astimezone(),
+            reverse=True,
+        )
+
+        if sorted_findings:
+            # only look at the most recent data
+            finding = sorted_findings[0]
+            if _skip_scan(
+                previous_finding=finding,
+                current_version_info=clamav_version_info,
+            ):
+                resource_name = f'{component.name}/{resource.name}'
+                logger.info(f'Skipping scan of {resource_name}.')
+                summary: dso.model.MalwareSummary = finding.data
+                if isinstance(resource.access, cm.OciAccess):
+                    resource_url = resource.access.imageReference
```

access-handling should be factored out
at the very least, raise for unknown resources, instead of simply ignoring/warning